### ISSUE TITLE
Scala 3 union types as error representation

### ISSUE THREAD

**Issue by msosnicki** [disneystreaming/smithy4s#703]
I had a look into error union types representation and compared the current state vs what it could look like if Scala 3 union type encoding was used. 
Below is the snippet that compares code generated for some simple operation (main errors section), and potential Scala 3 version.

```smithy
@http(method: "POST", uri: "/name/{name}", code: 200)
operation PostHello {
  input: Person,
  output: Greeting,
  errors: [BadRequest, InternalServerError]
}
```

```scala
object Old {
  sealed trait PostHelloError extends scala.Product with scala.Serializable {
    @inline final def widen: PostHelloError = this
  }
  object PostHelloError extends ShapeTag.Companion[PostHelloError] {
    val id: ShapeId = ShapeId("smithy4s.hello", "PostHelloError")

    val hints: Hints = Hints.empty

    case class BadRequestCase(badRequest: BadRequest) extends PostHelloError
    case class InternalServerErrorCase(internalServerError: InternalServerError)
        extends PostHelloError

    object BadRequestCase {
      val hints: Hints = Hints.empty
      val schema: Schema[BadRequestCase] = bijection(
        BadRequest.schema.addHints(hints),
        BadRequestCase(_),
        _.badRequest
      )
      val alt = schema.oneOf[PostHelloError]("BadRequest")
    }
    object InternalServerErrorCase {
      val hints: Hints = Hints.empty
      val schema: Schema[InternalServerErrorCase] = bijection(
        InternalServerError.schema.addHints(hints),
        InternalServerErrorCase(_),
        _.internalServerError
      )
      val alt = schema.oneOf[PostHelloError]("InternalServerError")
    }

    implicit val schema: UnionSchema[PostHelloError] = union(
      BadRequestCase.alt,
      InternalServerErrorCase.alt
    ) {
      case c: BadRequestCase          => BadRequestCase.alt(c)
      case c: InternalServerErrorCase => InternalServerErrorCase.alt(c)
    }
  }
}

object New {
  opaque type PostHelloError = BadRequest | InternalServerError
  object PostHelloError {
    val id: ShapeId = ShapeId("smithy4s.hello", "PostHelloError")

    def apply(error: BadRequest | InternalServerError): PostHelloError = error

    extension (e: PostHelloError)
      def value: BadRequest | InternalServerError = e

    val hints: Hints = Hints.empty
    implicit val schema: UnionSchema[PostHelloError] = {
      val badRequestAlt = BadRequest.schema.oneOf[PostHelloError]("BadRequest")
      val internalServerErrorAlt =
        InternalServerError.schema.oneOf[PostHelloError]("InternalServerError")
      union(badRequestAlt, internalServerErrorAlt) {
        case c: BadRequest          => badRequestAlt(c)
        case c: InternalServerError => internalServerErrorAlt(c)
      }
    }
  }
}

```

First, would union types encoding contribution be welcomed? 
And if so, what are your opinions about the above? There are design decisions to be made here, so wanted to discuss it a bit more (`opaque` vs `type alias` for example).

Thanks for all the work on smithy4s, it's a real joy to use and the internals are impressive

**Comment by Baccata** [comment]
Hi there, and thank you for opening the issue and for your appreciation, and best wishes for the new year 😄 

> First, would union types encoding contribution be welcomed? 

Yes, it would be very welcome, as an opt-in. It would have to be guided by means of metadata, prefixed with `smithy4s`. For an example, see https://disneystreaming.github.io/smithy4s/docs/codegen/customisation/unwrapping#default-rendering. 

So, something like `metadata smithy4sErrorsAsScalaUnions = true` would drive the decision the rendering. 

We're using metadata instead of build-tool settings as it saves us from having to duplicate the setting in 3 different contexts : sbt, mill and CLI. 

#### Regarding the schema : 

It looks good, kudos 

#### Regarding opaque types:  

Let's keep the type-alias there, if you're using ZIO or any kind of bifunctor, using opaque types would be detrimental to the UX of aggregating errors as you go. You'll however have to drop the `implicit` keyword from the `schema` value of this type (which really shouldn't be there for errors at all), as it will be useless to the implicit resolver. 

#### Regarding the implementation : 

The method below is that you should be amending. A combination of the `error` flag and a check for the metadata should drive the decision of whether to render scala 3 unions 

https://github.com/disneystreaming/smithy4s/blob/series/0.17/modules/codegen/src/smithy4s/codegen/internals/Renderer.scala#L583 

* The metadata could probably be propagated via the `CompilationUnit` construct [here](https://github.com/disneystreaming/smithy4s/blob/series/0.17/modules/codegen/src/smithy4s/codegen/internals/IR.scala#L36-L39). 
* You should probably create a `case class RenderingConfig(...)` with a `def decode(metadata: Map[String, Node]) : RenderingConfig` in the companion object (it's fine to throw exceptions in this method)
* Regarding testing, you can get inspiration from [this suite](https://github.com/disneystreaming/smithy4s/blob/series/0.17/modules/codegen/test/src/smithy4s/codegen/internals/DefaultRenderModeSpec.scala).

**Comment by msosnicki** [comment]
Ok, thanks for hints :) will try to come up with something

**Comment by msosnicki** [comment]
Looking into how error encoding is rendered, and how it shares code with general union encoding, I was wondering if there are any blockers to make it more broad, use Scala 3 unions everywhere where an `@adtMember` is not present?

**Comment by Baccata** [comment]
> Looking into how error encoding is rendered, and how it shares code with general union encoding, I was wondering if there are any blockers to make it more broad, use Scala 3 unions everywhere where an @adtMember is not present? 

That wouldn't work, because the following is totally valid and Scala unions would not capture the difference of semantics between the two members : 

``` 
union FooBar {
   foo: Integer
   bar: Integer
} 
```

**Comment by kubukoz** [comment]
Closed in #707
